exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/MarshalerInputStream.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp.trace;
+
+import com.google.common.io.ByteStreams;
+import com.google.protobuf.CodedOutputStream;
+import io.grpc.Drainable;
+import io.grpc.KnownLength;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import javax.annotation.Nullable;
+
+// Adapted from gRPC ProtoInputStream but using our Marshaller
+// https://github.com/grpc/grpc-java/blob/2c2ebaebd5a93acec92fbd2708faac582db99371/protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoInputStream.java
+final class MarshalerInputStream extends InputStream implements Drainable, KnownLength {
+  @Nullable private Marshaler message;
+  @Nullable private ByteArrayInputStream partial;
+
+  MarshalerInputStream(Marshaler message) {
+    this.message = message;
+  }
+
+  @Override
+  public int drainTo(OutputStream target) throws IOException {
+    int written;
+    if (message != null) {
+      written = message.getSerializedSize();
+      CodedOutputStream cos = CodedOutputStream.newInstance(target);
```

Comment:
FYI - There can be a huge performance swing in creating a CodedOutputStream without passing that serialized size in as the second parameter.  If you haven't benchmarked the difference, highly recommend you try it out.  I saw a large swing in results otherwise due to buffer resizing.